go/pserver/service.go

```diff
@@ -240,56 +260,46 @@ func (s *Service) GetParam(name string, parameter *Parameter) error {
 	return nil
 }
 
-// pserver save checkpoint
-func (s *Service) doCheckpoint() (err error) {
-	<-s.initialized
-	s.mu.Lock()
-	defer s.mu.Unlock()
+func traceTime(start time.Time, name string) {
+	elapsed := time.Since(start)
+	log.Infof("%s took %v", name, elapsed)
+}
+
+// checkpoint saves checkpoint to disk.
+//
+// checkpoint should be only called after the parameters are
+// initialized.
+func (s *Service) checkpoint() (err error) {
+	log.Infoln("Begin save checkpoint.")
+	defer traceTime(time.Now(), "save checkpoint")
```

Comment:
Let's keep this idea in mind. We still need to figure out the details like how to "merge" these logs together, because model evaluation is in trainer, but checkpoint is in pserver, and checkpoint overrides the previous checkpoint (so even if we know an old checkpoint is better, we can not recover it).
Another way is to let user implement model saving and evaluation by themselves in trainer alone.